java-squid/src/main/java/org/sonar/java/resolve/TypeResolutionHelper.java

```diff
@@ -0,0 +1,220 @@
+/*
+ * SonarQube Java
+ * Copyright (C) 2012 SonarSource
+ * sonarqube@googlegroups.com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.java.resolve;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.HashMultimap;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Lists;
+import com.google.common.collect.Maps;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Sets;
+import org.sonar.java.resolve.JavaType.ParametrizedTypeJavaType;
+import org.sonar.java.resolve.JavaType.TypeVariableJavaType;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
+
+public class TypeResolutionHelper {
+
+  private TypeResolutionHelper() {
+  }
+
+  /**
+   * Compute the "Least Upper Bound" ("lub", jls8 §4.10.4) of a list of type. is a shared supertype that is more specific than any
+   * other shared supertype (that is, no other shared supertype is a subtype of the least upper bound)
+   * @param types
+   * @return the least upper bound of the types
+   */
+  public static JavaType leastUpperBound(List<JavaType> types) {
+    Preconditions.checkArgument(types.size() > 1);
+
+    Multimap<JavaType, JavaType> supertypes = supertypes(types);
+    Multimap<JavaType, JavaType> erasedSupertypes = erasedSupertypes(supertypes);
+
+    Set<JavaType> erasedCandidates = intersectionOfErasedTypes(erasedSupertypes);
+    Set<JavaType> minimalErasedCandidates = minimalErasedCandidates(erasedCandidates);
+    if (minimalErasedCandidates.isEmpty()) {
+      return Symbols.unknownType;
+    }
+    Multimap<JavaType, ParametrizedTypeJavaType> relevants = relevant(minimalErasedCandidates, supertypes);
+    Map<JavaType, ParametrizedTypeJavaType> candidates = leastContainingParameterization(relevants);
+
+    List<JavaType> bests = best(minimalErasedCandidates, candidates);
+    if (bests.size() > 1) {
+      // FIXME should be the Best(W1) & ... & Best(Wr)
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1134)